### PR TITLE
bug-fix: Send safety for sync function

### PR DESCRIPTION
**Pull request by adrian-kong**
Adds send tests and attempts to fix some Send issues

Remaining issues:
1. 
```
thread 'main' panicked at 'Command esthri: command name `put` is duplicated', /Users/adriankong/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.5/src/builder/debug_asserts.rs:334:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### CODE REVIEW

File: crates/esthri/src/ops/sync.rs [hunk 1]
```diff
@@ -162,7 +162,7 @@ pub async fn sync(
                 key
             );
 
-            sync_remote_to_local(s3, &bucket, &key, &path, &filters, opts).await?;
+            // sync_remote_to_local(s3, &bucket, &key, &path, &filters, opts).await?;
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:165]: we are having issues with stream delete.
i think some streaming functionalities removes the Send.

current non thread safe methods are sync, 
think `AsRef<Path>` is not Send safe,
`delete_streaming()` is causing a big problem

File: crates/esthri/src/ops/sync.rs [hunk 2]
```diff
@@ -286,7 +286,7 @@ where
 {
     input_stream.map(move |params| async move {
         let (source_path, metadata) = params?;
-        let file_path: Box<dyn AsRef<Path>> = Box::new(source_path.clone());
+        let file_path = source_path.clone();
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:289]: changing to PathBuf to ensure send

File: crates/esthri/src/ops/sync.rs [hunk 3]
```diff
@@ -506,29 +495,26 @@ async fn sync_delete_local(
...
-        if let Some(s3_metadata) = s3_metadata {
-            let fs_metadata = fs::metadata(&path).await;
-            if let Err(err) = fs_metadata {
-                if err.kind() == ErrorKind::NotFound {
-                    let obj_path = String::from(key) + &s3_metadata.s3_suffix;
-                    Ok(Some(obj_path))
-                } else {
-                    Err(err.into())
+    let delete_paths_stream = stream.try_filter_map(move |obj| {
+        let key = key.to_string();
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:499]: this ensures Send is satisfied due to `&str` lifetimes

File: crates/esthri/src/ops/sync.rs [hunk 4]
```diff
@@ -643,46 +629,35 @@ async fn sync_across(
...
-            } else {
-                Ok(None)
-            }
-        },
-    );
+    let delete_paths_stream = bucket_stream.try_filter_map(move |(_, key, _)| {
+        let s3 = s3.clone();
+        let src_bucket = src_bucket.to_string();
+        let src_prefix = src_prefix.to_string();
+        let dst_prefix = dst_prefix.to_string();
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:646]: also satisfying Send trait

File: crates/esthri/src/ops/sync.rs (hunk 1 above)
Comment (adrian-kong) [reply crates/esthri/src/ops/sync.rs]: fixed

File: crates/esthri/src/ops/sync.rs [hunk 5]
```diff
@@ -302,11 +303,11 @@ where
                 // should be compressed
                 let (tmp, _) = compress_to_tempfile(&source_path).await?;
                 let local_etag = if metadata.is_some() {
-                    compute_etag(tmp.path()).await.map(Option::Some)
+                    compute_etag(tmp.path()).await.map(Some)
                 } else {
                     Ok(None)
                 };
-                (tmp.into_path(), source_path, local_etag)
+                (tmp.path().to_path_buf(), source_path, local_etag)
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:310]: this seems to be breaking

File: crates/esthri/src/ops/sync.rs (hunk 2 above)
Comment (adrian-kong) [reply crates/esthri/src/ops/sync.rs]: reverted to just `+ Send + Sync`

File: crates/esthri/src/ops/sync.rs [hunk 6]
```diff
@@ -286,12 +287,12 @@ where
 {
     input_stream.map(move |params| async move {
         let (source_path, metadata) = params?;
-        let file_path: Box<dyn AsRef<Path>> = Box::new(source_path.clone());
+        let file_path: Box<dyn AsRef<Path> + Send + Sync> = Box::new(source_path.clone());
```
Comment (adrian-kong) [inline crates/esthri/src/ops/sync.rs:290]: Adds `+Send+Sync` to fix passing sync() to thread.

**Review by silverjam — APPROVED** [review]

**Pull request merged**